### PR TITLE
Enable auto updates

### PR DESCRIPTION
**Pull request by alphadose**
The binary names in the [release list](https://github.com/gravitl/netmaker/releases/tag/v0.17.0) must adhere to `netclient-{platform}-{architecture}` which is already followed except for linux platforms, amd64 architectures and windows. Platform and architecture names must be mentioned explicitly for all built binaries.

Tested with https://github.com/alphadose/test-selfupdate

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -4,13 +4,35 @@ Copyright © 2022 Netmaker Team <info@netmaker.io>
...
+
+	"github.com/blang/semver"
 	"github.com/gravitl/netclient/cmd"
 	"github.com/gravitl/netclient/config"
+	"github.com/rhysd/go-github-selfupdate/selfupdate"
 )
 
 var version = "dev"
 
+func autoUpdate() {
```
Comment (Aceix) [inline main.go:18]: Let's put this in a different package, then import to main so that the GUI can later on use the it

**Review by Aceix — CHANGES_REQUESTED** [review]

File: functions/self_update.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/rhysd/go-github-selfupdate/selfupdate"
+)
+
+func SelfUpdate(currentVersion string) {
+	if currentVersion == "dev" {
+		return
+	}
+	semVer := strings.Replace(currentVersion, "v", "", -1)
+	v := semver.MustParse(semVer)
+	latest, err := selfupdate.UpdateSelf(v, "gravitl/netmaker")
```
Comment (Aceix) [inline functions/self_update.go:18]: Not sure but is this supposed to be "gravitl/netclient" rather?

Comment (alphadose) [reply functions/self_update.go:21]: netclient release binaries are present in the netmaker repo https://github.com/gravitl/netmaker/releases/tag/v0.17.0

File: main.go (hunk 1 above)
Comment (alphadose) [reply main.go]: done

File: functions/self_update.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package functions
+
+import (
+	"log"
+	"strings"
+
+	"github.com/blang/semver"
```
Comment (mattkasun) [inline functions/self_update.go:7]: https://github.com/hashicorp/go-version is more popular and more recently updated go version library.  Already used in netmaker branch netclient_refactor

File: functions/self_update.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+// SelfUpdate updates the netclient binary in place to the latest release available on GitHub
+// and reboots the daemon if update is successful
+// All binary names must adhere to the format `netclient-{platform}-{architecture}`
+func SelfUpdate(currentVersion string) {
+	if currentVersion == "dev" {
+		return
+	}
+	semVer := strings.Replace(currentVersion, "v", "", -1)
```
Comment (mattkasun) [inline functions/self_update.go:19]: is this going to handle a version like v0.16.3-testing?  
you probably want to do something like this
```
	trimmed := strings.TrimFunc(ver, func(r rune) bool {
		return !unicode.IsNumber(r)
	})
```

**Review by mattkasun — CHANGES_REQUESTED** [review]

File: functions/self_update.go (hunk 2 above)
Comment (mattkasun) [reply functions/self_update.go:21]: they won't be in the future and the binaries in the netmaker repo will not be compatible with the file structure and the servers that binaries from this repo communicate with

File: functions/self_update.go (hunk 3 above)
Comment (alphadose) [reply functions/self_update.go:7]: But the `go-github-selfupdate` library only works with `github.com/blang/semver`. Would need to fork the repo and refactor it to use `hashicorp/go-version`

File: functions/self_update.go (hunk 4 above)
Comment (alphadose) [reply functions/self_update.go]: done

File: functions/self_update.go (hunk 2 above)
Comment (alphadose) [reply functions/self_update.go]: done

File: functions/self_update.go (hunk 3 above)
Comment (mattkasun) [reply functions/self_update.go:8]: ok, didn't realize that

**Review by theguy951357 — APPROVED** [review]
made fork on my repo

created two forks off of feature_auto_update branch

chris_test1 and chris_test2

made sure both branches point to my repo and the version is v0.0.0 and v0.0.1 respectively

built binary off of chris_test1 "netclient_linux_amd64"

SCPed it to my mac and put it in a release v0.0.0

repeated with chris_test2 and put in release v0.0.1

wget the v0.0.0 binary and installed

[netclient-linux-amd64] 2023-01-06 18:37:32 Successfully updated to version 0.0.1 

[netclient-linux-amd64] 2023-01-06 18:37:32 Release notes:

  

2023/01/06 18:37:33 calling systemclt stop netclient




(there is a typo "systemclt" instead of systemctl. checked in code. it is just in the log. the actual command is correct.)

pass

**Review by Aceix — APPROVED** [review]

**Comment by alphadose** [comment]
@0xdcarns rebased with develop

**Review by 0xdcarns — APPROVED** [review]

**Pull request merged**